File upload and download

pandaponn opened the pull request:
Added methods for upload/download files to an FTP server along with db methods to save the successfully uploaded files into the appropriate tests and solutions tables. I have included the credentials for connecting to the local FTP server in local.properties but these credentials can be changed freely when you setup your own local FTP server.

pandaponn commented on bin/main/config/local.properties:
```diff
@@ -0,0 +1,11 @@
+FTP_PASS=password
+FTP_HOST=localhost
+FTP_PORT=21
+REMOTE_PATH=/upload123/
```
My bad, I must have duplicated it by accident whilst switching branches. Let me fix that!

HarveyRDonnelly commented on src/main/java/fworks/da/PostgresAccessManager.java:
```diff
@@ -138,4 +138,52 @@ public boolean saveNewUserInfo(URegisterDsRequestModel requestModel) {
     }
 
+    // Generate random alphanumeric ID
+    public static String randomIdGenerator(){
```
This looks great and will be very useful when implementing all other save methods for entities (where id generation is required).

HarveyRDonnelly commented on src/main/java/driver/FtpClient.java:
```diff
@@ -0,0 +1,159 @@
+            boolean done = ftpClient.storeFile(remoteFile, inputStream);
+            inputStream.close();
+            if (done) {
+                System.out.println("The file is uploaded successfully.");
```
I don't think print to the console is necessary here since we are not doing so for other operations in the software. Consistent logging across the project is a feature that we should consider for the future, but this is ok for now

pandaponn replied:
Do I remove the print statement or is it ok?

HarveyRDonnelly reviewed (changes requested):
This looks great, however, I believe that some changes are required in order to adhere to the CLEAN architecture. I believe the FtpClient class should be moved to the frameworks.daccess  package (fworks..da in old naming scheme) and renamed FtpAccessManager in order to be consistent with the PostgresAccessManager convention. A gateway interface should be created in the iadapters.gateways package (formerly ia.gateways) which is implemented by the FtpAccessManager. This gateway should implement the gateways defined for the relevant use cases (i.e. test and solution upload) so that the object can be injected into the use case layer using dependency inversion.

pandaponn commented on src/main/java/ia/gateways/FileAccessGateway.java:
```diff
@@ -0,0 +1,5 @@
+package ia.gateways;
+
+public interface FileAccessGateway {
```
I believe we have to merge Nicholas' latest PR before I can extend the necessary file access gateways for the use cases? @HarveyRDonnelly

HarveyRDonnelly replied:
Yes but that may be added in a future PR.

pandaponn commented:
Updated FtpClient to clean architecture by implementing FtpAccessManager and gateway FileAccessGateway.

Included javadoc for FtpAccessManager.

Added db queries to get test and solution file names. Changed snake case to camel case. @HarveyRDonnelly

HarveyRDonnelly reviewed (approved):
This looks great @pandaponn ! After implementing the requested changes, the code now adheres to the CLEAN architecture. JavaDoc and use case gateway implementations may be added in a future PR.

HarveyRDonnelly reviewed (approved):
Merge conflicts in build.gradle resolved and PR is ready to merge.
